docs-internal/multipart.md

```diff
@@ -0,0 +1,328 @@
+# Helidon Multipart Support Proposal
+
+Provide an API for the reactive webserver that enables processing multipart requests and returning multipart responses.
+
+## Proposal
+
+Support 2 models: buffered and reactive.
+
+With the buffered model the request content is fully decoded in memory and all the parts are handed-over at the same time.
+This model enables simple use-cases with small size parts.
+
+With the reactive model the request content is decoded for each chunk and the user can subscribe to the stream of chunks for each part.
+This model enables advanced streaming use-cases with large size parts.
+
+Supporting multipart in the webserver means adding support for a new media type. This is done by adding a new entity type with readers and writers.
+
+Each of the 2 models above will be represented by different entity types: `BufferedMultiPart` and `MultiPart`.
+
+```java
+// buffered model
+request.content().as(BufferedMultiPart.class).thenAccept(/* ... */);
+```
+
+```java
+// reactive model
+request.content().as(MultiPart.class).thenAccept(/* ... */);
+```
+
+## Examples
+
+Let's illustrate with code examples first.
+
+### Enable Multipart Support
+
+The readers and writers for the multipart entity types will be registered by a service, similar to how other media types are supported in the webserver.
+All the examples below assume that the MultiPartSupport is registered as follow:
+
+```java
+Routing.builder()
+    .register(MultiPartSupport.create())
+    // user defined handlers...
+    .build());
+```
+
+### Process a request with the buffered model
+
+The following example consumes each part content as a `JsonObject`:
+
+```java
+request.content().as(BufferedMultiPart.class).thenAccept(mp -> {
```

Comment:
That's a fair point.

Having a single top level entity for both buffered and reactive would require providing callbacks for the user to consume the parts and return a response.
This approach would make both models equally verbose.

The main advantage of having 2 different entity types is that it makes the simple case much easier. There is a single callback: the CompletionStage<BufferedMultiPart>, and then every can be written synchronously.

Reply:
Oh, yes. I'm completely on-board with making the buffered case as easy as possible. Hence my earlier note (without, sadly, a suggestion), for an alternative way for the developer to trigger the buffered behavior.

If we had just the one entity type `MultiPart` then could we do something like this to simplify the buffered case: 
```java
request.content().as(MultiPart.class).buffered().thenAccept(bufferedMP -> {...})
```
and the rest of the example could be exactly the same for the buffered case as in the original proposal.

I know it's not much different from the current proposal with the two entity types, but it does separate the inherent nature of the content from how the developer wishes to process it.

Reply:
There is an issue with your example: the `as` method returns a `CompletionStage<T>`.
It would look rather like this:

```java
request.content().as(MultiPart.class).thenAccept((mp) ->{
    mp.buffered.thenAccept((bufferedMp) -> {
        // ...
    });
});
```
Even with the additional callback this may be a valid option.